### PR TITLE
DRILL-5425: Support HTTP Kerberos auth using SPNEGO

### PR DESCRIPTION
**Pull request by sohami**

### CODE REVIEW

**Comment by priteshm** [comment]
@arina-ielchiieva can you please review this?

**Comment by arina-ielchiieva** [comment]
@sohami could you please update pull request header to include Jira title? It would be just easier to identify it among others. Thanks.

File: exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/auth/SpnegoUtil.java [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+      if (!UserGroupInformation.isSecurityEnabled()) {
+        final Configuration newConfig = new Configuration();
+        newConfig.set(CommonConfigurationKeys.HADOOP_SECURITY_AUTHENTICATION,
+            UserGroupInformation.AuthenticationMethod.KERBEROS.toString());
+
+        UserGroupInformation.setConfiguration(newConfig);
+        ugi = UserGroupInformation.loginUserFromKeytabAndReturnUGI(principal, keytab);
+
+        // Reset the original configuration for static UGI
+        UserGroupInformation.setConfiguration(new Configuration());
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/auth/SpnegoUtil.java]: Is there any possible concurrency issues when calling `UserGroupInformation.setConfiguration`? What if two users try to `loginAndReturnUgi` at the same time? Even if users will have different `SpnegoUtil` instances in `UserGroupInformation` we operate at the static level.

File: exec/java-exec/src/test/java/org/apache/drill/exec/rpc/security/KerberosHelper.java [hunk 2]
```diff
@@ -43,16 +45,20 @@
   public String SERVER_PRINCIPAL;
   private final String testName;
 
-  private File keytabDir;
+  public File keytabDir;
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/test/java/org/apache/drill/exec/rpc/security/KerberosHelper.java]: Generally I am not in favor of public variables (even in test classes helpers). Better to provide getters if needed.

File: exec/java-exec/src/test/java/org/apache/drill/exec/server/TestSpnegoAuthentication.java [hunk 3]
```diff
@@ -0,0 +1,597 @@
...
+    assertTrue(!securityProvider.isSpnegoEnabled());
+  }
+
+  /**
+   * Validate behavior of {@link SpnegoUtil} class when provided with different configuration's for SPNEGO via
+   * DrillConfig
+   * @throws Exception
+   */
+  @Test
+  public void testSpnegoUtil() throws Exception {
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/test/java/org/apache/drill/exec/server/TestSpnegoAuthentication.java]: Test for spnego util can be factored out in different class plus test can be split into four.

File: exec/java-exec/src/test/java/org/apache/drill/exec/server/TestSpnegoAuthentication.java [hunk 4]
```diff
@@ -0,0 +1,597 @@
...
+    assertTrue(user.isUserInRole("authenticated", null));
+  }
+
+  /**
+   * Validate {@link DrillSpnegoAuthenticator} with request of different nature and from authenticated and
+   * unauthenticated session.
+   * @throws Exception
+   */
+  @Test
+  public void testDrillSpnegoAuthenticator() throws Exception {
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/test/java/org/apache/drill/exec/server/TestSpnegoAuthentication.java]: The same here, test can be factored out in the separate class dedicated `DrillSpnegoAuthenticator` testing and split into several.

**Comment by arina-ielchiieva** [comment]
@sohami thanks for making the changes. +1, LGTM.

**Pull request closed without merging**